### DISCUSSION TITLE
Concurrent schedules - how to make a stimulus appear independently during trials of another task?

### DISCUSSION THREAD

**Discussion by mbeasle2** [jspsych/jsPsych#3130] [General]
I'm looking to design an experiment where there are essentially two independent tasks running at the same time and a participant will need to multi-task. I found jsPsych and I am learning, but I didn't want to get too far unless I knew this was possible, so forgive me if this is a repeat question; I did a check and found something kind of similar, but not quite answering my question.

My participants will be doing *something*, let's say, they'll be typing in responses to math questions. Each time they hit 'enter' and new math question will appear. Correct answers give points based on how quickly they answer.

While they're doing this, every 15-45s, I need another stimulus to appear on screen. When it appears, they will make another key response (say, 'space') and make it go away. If they do not press in time, they lose points. 

My biggest concern is how I would code each of these variables to be presented simultaneously, so that perhaps they take a long time on one math question and must respond to the stimulus several times during one math trial, but then during another math trial they may be so quick that they will not need to respond at all. 

Thank you so much!

**Comment by jodeleeuw** [comment]
Hi @mbeasle2,

This is a tricky use case for jsPsych. It's definitely possible, but there's no obvious implementation given the way that jsPsych experiments are structured.

One option would be to create a new plugin to handle this. In the plugin code you'd have full control over the display and you could basically build both tasks within the plugin. If I were building this I'd probably focus on the primary task, e.g., solving the math problem and design the plugin so that one trial is one math problem. To handle the secondary task, I'd probably add parameters like `first_display_time` to specify how long after the start of the trial the stimulus should appear and `repeat_display_time` to specify how often it should appear. To create the full timeline, I'd use dynamic parameters to set `first_display_time` so that I could use information from the end of the previous trial to figure out when the next stimulus should appear.

A second option might be to build [an extension](https://www.jspsych.org/7.3/overview/extensions/). The extension would handle the secondary task, and then this could be layered on top of any plugin. I think this could also have its own internal state to keep track of the timing. This is probably the better way to do it, but there is less documentation for extension writing out there.

A third option would be to implement everything inside the `on_load` and `on_finish` functions for each trial. This wouldn't require any new plugins or extensions, but it would require some "hacky" solutions to get the data from the reaction to the stimulus into the jsPsych data. I don't think I'd advise going this route, but I could provide a little more direction on how this might work if this is the way you'd like to go.

**Reply by mbeasle2** [reply]
Hi Josh, 

I have code for two different tasks now, and I want to combine them. I am going to try to use your plugin method. So, I edited the survey text plugin to make a timed math problem task with a score counter. Should I just add in the parameters like you suggested from my RT task? Thank you!
